### ISSUE TITLE
Debug and fix unreliable QtDims tests

### ISSUE THREAD

**Issue by andy-sweet** [napari/napari#5443]
## 🧰 Task

Recently, a few tests defined in `test_qt_dims` have been failing fairly frequently (but not consistently) for non-obvious reasons. 

`test_play_button` has been [aborting sometimes](https://github.com/napari/napari/actions/runs/3761999271/jobs/6394264970#step:8:310) on main and PRs. This is likely due to #5414, though that itself was an attempt to make this test less flaky, so reverting that PR will only help a little.

`test_not_playing_after_ndim_changes` has been [hanging sometimes](https://github.com/napari/napari/actions/runs/3751034702/jobs/6372375860). Maybe more after #5373, though I'm not sure.

The task here is to understand why these tests are failing and either fix them or fix the underlying implementation. Removing the tests with a good explanation would also be a valid fix.

**Comment by Carreau** [comment]
I can suggest as well to move them to a different item matrix, so that when hanging/failing it's easy to run just those.

**Comment by Carreau** [comment]
I can't reproduce the failure locally, but after running those test a bunch of time I got this warning:

```
napari/_qt/widgets/_tests/test_qt_dims.py::test_not_playing_after_ndim_changes
  ~/miniforge3/envs/nap6/lib/python3.11/site-packages/superqt/utils/_ensure_thread.py:42: RuntimeWarning: PySide6 Warning: Skipping callback call set_result17a4f5b10104817920(PyObject) because the callback object is being destructed.
    self.finished.emit(res)
```

Maybe this will help figuring it out ?

**Comment by Carreau** [comment]
And the thing this refer's to seem to be `self._animation_thread` of QtDims.

**Comment by Czaki** [comment]
Hm. There is a need to rethink QtDims implementation...

**Comment by Carreau** [comment]
Ha ! #5466 make the test fail reliably ! So it seem that something is keeping too many references to the thread.

```
$ pytest  --cov-report=xml --cov=napari --ignore tools --maxfail=5 napari/_qt/widgets/_tests/test_qt_dims.py::test_play_button
========================================================================== test session starts ===========================================================================
platform darwin -- Python 3.11.0, pytest-7.2.0, pluggy-1.0.0
PyQt5 5.15.7 -- Qt runtime 5.15.6 -- Qt compiled 5.15.6
rootdir: /Users/bussonniermatthias/dev/napari, configfile: pyproject.toml
plugins: napari-0.4.17rc4.dev134+g600188672.d20221222, qt-4.2.0, napari-plugin-engine-0.2.0, npe2-0.6.1, cov-4.0.0, hypothesis-6.61.0
collected 1 item

napari/_qt/widgets/_tests/test_qt_dims.py Fatal Python error: Aborted

Current thread 0x0000000102dd4580 (most recent call first):
  File "~/dev/napari/napari/_qt/widgets/qt_dims.py", line 312 in cleaned_worker
  File ".../miniforge3/envs/nap6/lib/python3.11/site-packages/pytestqt/qt_compat.py", line 162 in exec
  File ".../miniforge3/envs/nap6/lib/python3.11/site-packages/pytestqt/wait_signal.py", line 51 in wait
```

Which point to :

```
    @Slot()
    def cleaned_worker(self):
        self._animation_thread = None # <- here.
        self._animation_worker = None
        self.enable_play()
```

Putting a `gc.collect()`, just before seem to fix it

**Comment by Czaki** [comment]
but this log means that `cleande_worker` is called after `QtDims` is destroyed.

**Comment by Carreau** [comment]
> but this log means that `cleaned_worker` is called after `QtDims` is destroyed.

Yeah, I'm confused as well, I'm thinking of a closure or something else, that keeps something around that should not be. I've tried to convert some reference to the thread to weak refs, and it crashes earlier.

**Comment by Czaki** [comment]
There is a need to keep the hard reference to a thread in one place, otherwise, gc will lead to a crash.

**Comment by Carreau** [comment]
Yes, and we need to try to keep a single reference and make sure we manipulate this thread only from this single reference.

**Comment by Czaki** [comment]
I have made some investigation and maybe solve this issue in #6594
